### PR TITLE
MDEV-21910 Deadlock between BF abort and manual KILL command

### PR DESCRIPTION
**Pull request by sjaakola**
Added THD::wsrep_killed flag to synchronize who can kill the victim
This flag is set both when BF is called for from innodb and by KILL command
Either path of victim kill will bail out if victim's wsrep_killed is already
set to avoid mutex conflicts with the other aborter execution.

A new test case was added in galera.galera_bf_kill.test for scenario where
wsrep applier thread and manual KILL command try to kill same idle victim

### CODE REVIEW

**Comment by sjaakola** [comment]
fleshed out a better commit message for the PR

File: sql/service_wsrep.cc [hunk 1]
```diff
@@ -302,3 +320,13 @@ extern "C" void wsrep_commit_ordered(THD *thd)
...
+
+extern "C" bool wsrep_thd_set_wsrep_killed(THD *thd)
+{
+  if (thd->wsrep_killed)
+  {
+    return true;
+  }
+  thd->wsrep_killed = true;
+  return false;
+}
```
Comment (dr-m) [inline sql/service_wsrep.cc:332]: In `wsrep_thd_bf_abort()`, `thd->wsrep_killed` is protected by `thd->LOCK_thd_data`. Can we protect it also here? In the current state, we can have a race condition between multiple `wsrep_thd_set_wsrep_killed()` executions.

If using a mutex is not possible, could we at least use `Atomic_wrapper` for `wsrep_killed`, and perform the test-and-set as an atomic operation?

The assignment operator in MariaDB code should not be preceded by a space:
```c++
thd->wsrep_killed= true;
```

File: sql/sql_class.h [hunk 2]
```diff
@@ -4854,6 +4854,7 @@ class THD: public THD_count, /* this must be first */
     table updates from being replicated to other nodes via galera replication.
   */
   bool                      wsrep_ignore_table;
+  bool                      wsrep_killed;
```
Comment (dr-m) [inline sql/sql_class.h:4857]: A comment that describes the purpose of the flag and the locking rules is missing.

File: sql/sql_class.cc [hunk 3]
```diff
@@ -2126,6 +2128,9 @@ void THD::reset_killed()
     killed_err= 0;
     mysql_mutex_unlock(&LOCK_thd_kill);
   }
+#ifdef WITH_WSREP
+  wsrep_killed = false;
+#endif /* WITH_WSREP */
```
Comment (dr-m) [inline sql/sql_class.cc:2133]: Apparently, no mutex is protecting `wsrep_killed` here.

File: sql/sql_parse.cc [hunk 4]
```diff
@@ -9080,8 +9079,25 @@ kill_one_thread(THD *thd, longlong id, killed_state kill_signal, killed_type typ
...
+      {
+#endif /* WITH_WSREP */
+      WSREP_DEBUG("kill_one_thread %llu, victim: %llu wsrep_killed %d by signal %d",
+                  thd->thread_id, id, tmp->wsrep_killed, kill_signal);
       tmp->awake_no_mutex(kill_signal);
+      WSREP_DEBUG("victim: %llu taken care of", id);
       error=0;
+#ifdef WITH_WSREP
+      }
+#endif /* WITH_WSREP */
```
Comment (dr-m) [inline sql/sql_parse.cc:9100]: Indent with TABs, not spaces. What is protecting `tmp->wsrep_killed` here?

Please avoid excessive `#ifdef`:
```c++
#ifdef WITH_WSREP
if (…)
{
…
}
else
#endif
{
…
}
```

File: storage/innobase/handler/ha_innodb.cc [hunk 5]
```diff
@@ -18687,10 +18687,17 @@ wsrep_innobase_kill_one_trx(
 
 	/* Mark transaction as a victim for Galera abort */
 	victim_trx->lock.was_chosen_as_wsrep_victim= true;
+	if (wsrep_thd_set_wsrep_killed(thd))
+	{
+	  WSREP_DEBUG("innodb kill transaction skipped due to wsrep_killed");
+	  wsrep_thd_UNLOCK(thd);
+	  DBUG_RETURN(0);
+	}
```
Comment (dr-m) [inline storage/innobase/handler/ha_innodb.cc:18695]: What protects the `thd` state in the two new function calls?

File: sql/service_wsrep.cc (hunk 1 above)
Comment (janlindstrom) [reply sql/service_wsrep.cc:332]: I have not yet thoroughly reviewed this but if we at this point own lock_sys->mutex and trx->mutex(victim) LOCK_thd_data is exactly something we should not take, it just moves deadlock on different location.

File: sql/sql_parse.cc (hunk 4 above)
Comment (janlindstrom) [reply sql/sql_parse.cc:9100]: find_thread_by_id() acquires LOCK_thd_data before this

File: sql/sql_class.cc (hunk 3 above)
Comment (janlindstrom) [reply sql/sql_class.cc:2133]: LOCK_thd_kill mutex would protect from concurrent awake() call, does this really need to be outside of that if and mutex protection ?

File: sql/service_wsrep.cc (hunk 1 above)
Comment (janlindstrom) [reply sql/service_wsrep.cc:332]: Only place where this is called already has locked LOCK_thd_data

File: sql/sql_class.cc (hunk 3 above)
Comment (janlindstrom) [reply sql/sql_class.cc:2133]: I did not analyze all code paths to here so not sure if LOCK_thd_data is already locked, I would recommend either

mysql_mutex_assert_not_owner(&victim_thd->LOCK_thd_data);
mysql_mutex_lock(&LOCK_thd_kill);
wsrep_killed= false;
mysql_mutex_unlock(&LOCK_thd_kill);

or
mysql_mutex_assert_owner(&victim_thd->LOCK_thd_data);
wsrep_killed= false;

File: sql/sql_class.h (hunk 2 above)
Comment (janlindstrom) [reply sql/sql_class.h:4857]: It seems that we could protect it with LOCK_thd_data mutex

**Comment by dr-m** [comment]
Apparently something like this was applied to 10.4 in 956f21c3b0beba89835e8d24fde4b67b589ab142 and to 10.5 in 7bffe468b239645d2f27d1d5625cb9c914ae994d.

File: sql/sql_parse.cc [hunk 6]
```diff
@@ -9080,8 +9079,25 @@ kill_one_thread(THD *thd, longlong id, killed_state kill_signal, killed_type typ
         thd->security_ctx->user_matches(tmp->security_ctx))
 #endif /* WITH_WSREP */
     {
+#ifdef WITH_WSREP
+      DEBUG_SYNC(thd, "before_awake_no_mutex");
+      if (tmp->wsrep_killed)
+      {
+        /* victim is in hit list already, bail out */
+	WSREP_DEBUG("victim has wsrep_killed set, skipping awake()");
+	error = 0;
```
[the remark this answers is not in the document]
Comment (dr-m) [reply sql/sql_parse.cc:9090]: I noticed exactly the same thing in @sysprg’s merge of this to 10.4 (956f21c3b0beba89835e8d24fde4b67b589ab142).

**Pull request closed without merging**